Source/Fluids/WarpXFluidContainer.cpp

```diff
@@ -0,0 +1,743 @@
+    // - Compute the cell slopes
+        // - ave function required ( minmod )
+    // - Predict Q at dt/2
+    // - Predict Q at cell edges using the cell slopes
+    // - Update Q at t + dt
+    // - Update N, NU
+
+    // Temporary velocities
+    amrex::MultiFab tmp_Vx(N[lev]->boxArray(), N[lev]->DistributionMap(), 1, 1);
+    amrex::MultiFab tmp_Vy(N[lev]->boxArray(), N[lev]->DistributionMap(), 1, 1);
+    amrex::MultiFab tmp_Vz(N[lev]->boxArray(), N[lev]->DistributionMap(), 1, 1);
+
+    // Temporary Half-step values
+    amrex::MultiFab tmp_Q_minus_x(N[lev]->boxArray(), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_x(N[lev]->boxArray(), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_minus_y(N[lev]->boxArray(), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_y(N[lev]->boxArray(), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_minus_z(N[lev]->boxArray(), N[lev]->DistributionMap(), 4, 1);
+    amrex::MultiFab tmp_Q_plus_z(N[lev]->boxArray(), N[lev]->DistributionMap(), 4, 1);
+
+    // Advection push
+    #ifdef AMREX_USE_OMP
+    #pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+    #endif
+    for (MFIter mfi(*N[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+        amrex::Box const &tile_box = mfi.tilebox(N[lev]->ixType().toIntVect());
+        amrex::Array4<Real> const &N_arr = N[lev]->array(mfi);
+        amrex::Array4<Real> const &NUx_arr = NU[lev][0]->array(mfi);
+        amrex::Array4<Real> const &NUy_arr = NU[lev][1]->array(mfi);
+        amrex::Array4<Real> const &NUz_arr = NU[lev][2]->array(mfi);
+
+        amrex::Array4<amrex::Real> Vx = tmp_Vx.array(mfi);
+        amrex::Array4<amrex::Real> Vy = tmp_Vy.array(mfi);
+        amrex::Array4<amrex::Real> Vz = tmp_Vz.array(mfi);
+
+        amrex::Array4<amrex::Real> Q_minus_x = tmp_Q_minus_x.array(mfi);
+        amrex::Array4<amrex::Real> Q_plus_x = tmp_Q_plus_x.array(mfi);
+
+        amrex::Array4<amrex::Real> Q_minus_y = tmp_Q_minus_y.array(mfi);
+        amrex::Array4<amrex::Real> Q_plus_y = tmp_Q_plus_y.array(mfi);
+
+        amrex::Array4<amrex::Real> Q_minus_z = tmp_Q_minus_z.array(mfi);
+        amrex::Array4<amrex::Real> Q_plus_z = tmp_Q_plus_z.array(mfi);
+
+
+        amrex::ParallelFor(tile_box,
+            [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept
+            {
+
+                // - Grab local Uz Uy Ux gamma
+                // Isolate U from NU
+                auto Ux = (NUx_arr(i, j, k) / N_arr(i,j,k));
+                auto Uy = (NUy_arr(i, j, k) / N_arr(i,j,k));
+                auto Uz = (NUz_arr(i, j, k) / N_arr(i,j,k));
+                auto gamma = sqrt(1.0 + (Ux*Ux + Uy*Uy + Uz*Uz)/(clight*clight) );
+                auto a = (clight*clight*clight*clight)+(2.0*(Uz*Uz)+2.0*(Uy*Uy)+
+                2.0*(Ux*Ux))*(clight*clight)+(Uz*Uz*Uz*Uz)+(2.0*(Uy*Uy)+2.0*(Ux*Ux))
+                *(Uz*Uz)+(Uy*Uy*Uy*Uy)+2.0*(Ux*Ux)*(Uy*Uy)+(Ux*Ux*Ux*Ux);
```

Comment:
If I am not mistaken, this is simply `(c*gamma)^4`, which can be written
```
auto a = amrex::Math::powi<4>( clight * gamma );
```

Reply:
It is, let me replace that